contribute/design/1.x/csi/20190805-csi-zfspv-volume-provisioning.md

```diff
@@ -0,0 +1,139 @@
+already running on the node and the higher order applications should be
+able to conume this volume.
+
+## Proposal
+
+### User Stories
+
+#### Create a volume
+I should be able to provide a volume that can be consumed by my application.
+This volume should get created dynamically during application creation time
+and the provision should happen from the ZFS pool which is already running
+on the local nodes.
+
+#### Delete the volume
+I should be able to delete the volume that was being consumed by my application.
+This volume should get deleted only when the application is deleted and it should
+be cleand up from the ZFS pool running on the node.
+
+### Implementation Details
+
+#### Volume Creation Workflow
+- CSI driver will handle CSI request for volume create
+- CSI driver will read the request parameters and create following resources:
+  - ZfsVolumeCR _(Kubernetes custom resource)_
+- ZfsVolumeCR will be watched for the property change
+
+#### Volume Deletion Workflow
+- CSI driver will handle CSI request for volume delete
+- CSI driver will read the request parameters and delete corresponding ZFSPV volume
+resource
+
+#### High Level Design
+- user will setup all the node and setup the ZFS pool on each of those nodes.
+- user will deploy deblow sample storage class where we will get input of all the needed zfs properties.
+The storage class will have allowedTopologies and poolname which will tell us that pool is available on
+those nodes and it can pick any of that node to schedule the application pod.
+
+```yaml
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: openebs-zfspv
+provisioner: openebs.io/openebs-csi
+allowVolumeExpansion: true
+parameters:
+  blocksize: "4k"
+  compression: "on"
+  dedup: "on"
+  thinprovision: "yes"
+  poolname: "zfspv-pool"
+volumeBindingMode: WaitForFirstConsumer
+allowedTopologies:
+- matchLabelExpressions:
+  - key: kubernetes.io/hostname
+    values:
+    - csi-ubuntu-16
+```
+
+- user will deploy a PVC using above storage class
+
+```yaml
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: demo-zfspv-vol-claim
+spec:
+  storageClassName: openebs-zfspv
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 5G
+```
+## Workflow
+
+### 1. CSI create volume
+At CSI when we get a volume create request, we will create a PV object.
+The actual volume will be create at the Node publish time. We will have
+StorageClass at this point. We will create a ZfsVolumeCR with all the
+volume properties that are present in the custom resource.
+
+
+### 2. CSI Node Publish
+
+When kubernetes schedules the application pod according the Topology mentioned in
+the StorageClass, The CSI will get a NodePublish event. It will get all the volume
+properties from ZfsVolumeCR custom resource and it will fire zfs create command
+and creates the volume. It will get the pool name from the storage class and create
```

Comment:
Let CSI take in all the values/config/properties from ZFSVolume object.
Let us avoid making use of StorageClass at this stage.

Reply:
corrected it. we are taking that from CR only.